client/network/src/service.rs

```diff
@@ -716,34 +767,38 @@ impl<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> Stream for Ne
 					self.network_service.add_known_address(peer_id, addr),
 				ServerToWorkerMsg::SyncFork(peer_ids, hash, number) =>
 					self.network_service.user_protocol_mut().set_sync_fork_request(peer_ids, &hash, number),
+				ServerToWorkerMsg::EventsStream(sender) =>
+					self.events_streams.push(sender),
+				ServerToWorkerMsg::WriteNotif { message, proto_name, target } =>
+					self.network_service.user_protocol_mut().write_notif(target, proto_name, message),
+				ServerToWorkerMsg::RegisterNotifProtocol { proto_name, engine_id, handshake } =>
+					self.network_service.user_protocol_mut()
+						.register_notif_protocol(proto_name, engine_id, handshake),
 			}
 		}
 
 		loop {
 			// Process the next action coming from the network.
 			let poll_value = self.network_service.poll();
 
-			let outcome = match poll_value {
+			match poll_value {
 				Ok(Async::NotReady) => break,
-				Ok(Async::Ready(Some(BehaviourOut::SubstrateAction(outcome)))) => outcome,
-				Ok(Async::Ready(Some(BehaviourOut::Dht(ev)))) =>
-					return Ok(Async::Ready(Some(Event::Dht(ev)))),
-				Ok(Async::Ready(None)) => CustomMessageOutcome::None,
+				Ok(Async::Ready(Some(BehaviourOut::BlockImport(origin, blocks)))) =>
+					self.import_queue.import_blocks(origin, blocks),
+				Ok(Async::Ready(Some(BehaviourOut::JustificationImport(origin, hash, nb, justification)))) =>
+					self.import_queue.import_justification(origin, hash, nb, justification),
+				Ok(Async::Ready(Some(BehaviourOut::FinalityProofImport(origin, hash, nb, proof)))) =>
+					self.import_queue.import_finality_proof(origin, hash, nb, proof),
+				Ok(Async::Ready(Some(BehaviourOut::Event(ev)))) => {
+					self.events_streams.retain(|sender| sender.unbounded_send(ev.clone()).is_ok());
```

Comment:
We expect `unbounded_send` not to fail in many other places. Instead of silently dropping the events stream here, how about logging the failure?

Reply:
It can only fail if the receiving side gets dropped. IMO dropping the receiver is a clean and legitimate way to indicate that we're no longer interested in events.

Reply:
That makes sense. Thanks!